src/test/java/com/google/sps/servlets/AddClassTATest.java

```diff
@@ -0,0 +1,285 @@
+
+    List<Key> taClassList = Arrays.asList(init.getKey());
+
+    user.setProperty("userEmail", "test@google.com");
+    user.setProperty("registeredClasses", Collections.emptyList());
+    user.setProperty("taClasses", taClassList);
+    user.setProperty("ownedClasses", Collections.emptyList());
+
+    datastore.put(user);
+
+    // Create examples for the TA email and class code
+    when(httpRequest.getParameter("taEmail")).thenReturn("test@google.com");
+    when(httpRequest.getParameter("classCode")).thenReturn("exampleClassCode");
+
+    addTA.doPost(httpRequest, httpResponse);
+
+    Query query = new Query("User");
+    PreparedQuery results = datastore.prepare(query);
+
+    // Add the class key to the user's TA classes list
+    for (Entity entity : results.asIterable()) {
+      if (entity.getProperty("userEmail") == "test@google.com") {
+        List<Key> taClasses = (List<Key>) entity.getProperty("taClasses");
+        assertTrue(taClasses.contains(KeyFactory.stringToKey("exampleClassCode")));
+        assertTrue(taClasses.contains(init.getKey()));
+        assertTrue(taClasses.size() == 2);
+      }
+    }
+  }
+
+  @Test
+  // Verify that duplicate classes don't get added
+  public void preventDuplicates() throws Exception {
+
+    // Create some classes
+    Entity init = new Entity("Class");
+
+    init.setProperty("owner", "ownerID");
+    init.setProperty("name", "testClass");
+    init.setProperty("beingHelped", new EmbeddedEntity());
+    init.setProperty("studentQueue", Collections.emptyList());
+    init.setProperty("taList", Collections.emptyList());
+
+    Entity init2 = new Entity("Class");
+
+    init2.setProperty("owner", "ownerID2");
+    init2.setProperty("name", "testClass2");
+    init2.setProperty("beingHelped", new EmbeddedEntity());
+    init2.setProperty("studentQueue", Collections.emptyList());
+    init2.setProperty("taList", Collections.emptyList());
+
+    datastore.put(init);
+
+    // Initialize a TA user
+    Entity user = new Entity("User");
+
+    List<Key> taClassList = Arrays.asList(init.getKey(), init2.getKey());
+
+    user.setProperty("userEmail", "testTA@google.com");
+    user.setProperty("registeredClasses", Collections.emptyList());
+    user.setProperty("taClasses", taClassList);
+    user.setProperty("ownedClasses", Collections.emptyList());
+
+    datastore.put(user);
+
+    // Create examples for the TA email and class code
+    when(httpRequest.getParameter("taEmail")).thenReturn("testTA@google.com");
+    when(httpRequest.getParameter("classCode")).thenReturn(KeyFactory.keyToString(init.getKey()));
+
+    addTA.doPost(httpRequest, httpResponse);
+
+    verify(httpResponse).sendError(HttpServletResponse.SC_FORBIDDEN);
+
+    Query query = new Query("User");
+    PreparedQuery results = datastore.prepare(query);
+
+    // Verify list of classes for TA is unchanged
+    for (Entity entity : results.asIterable()) {
+      if (entity.getProperty("userEmail") == "testTA@google.com") {
```

Comment:
This is not actually testing what you are trying to test. The `==` operator tests for address equivalence (The two strings are in the location in memory). Instead you should use `.equals` which tests for structural equivalence (The two strings have the same contents). This tests appears to pass because this if statement is never true. 

An even better way to handle this is to use a `Filter` on the query so that you only get one result back.

Reply:
Changed to single filter